File: backend/deriv_ws.py
# backend/deriv_ws.py

import json
import websocket
from config import DERIV_APP_ID, DERIV_WS_URL

def subscribe_ticks(symbol):
    ws = websocket.WebSocket()
    ws.connect(f"{DERIV_WS_URL}?app_id={DERIV_APP_ID}")

    msg = {
        "ticks": symbol,
        "subscribe": 1
    }
    ws.send(json.dumps(msg))

    while True:
        data = json.loads(ws.recv())
        yield data

File: static/core/deriv_ws.js
/* =========================================================
   FILE: deriv_ws.js
   PURPOSE: Live ticks from Deriv WebSocket (VIX, Boom/Crash)
   ENGINE: ChequeWS™
========================================================= */

window.createDerivStream = function(symbol = "R_75") {

    const ws = new WebSocket("wss://ws.binaryws.com/websockets/v3?app_id=1089");

    let listeners = [];

    ws.onopen = () => {
        ws.send(JSON.stringify({
            ticks: symbol,
            subscribe: 1
        }));
    };

    ws.onmessage = (msg) => {
        const data = JSON.parse(msg.data);

        if (data.tick) {
            listeners.forEach(cb => cb({
                symbol: data.tick.symbol,
                quote: data.tick.quote,
                epoch: data.tick.epoch
            }));
        }
    };

    ws.onerror = () => console.warn("Deriv WS error.");
    ws.onclose = () => console.log("Deriv WS closed.");

    return {
        onTick(callback) {
            listeners.push(callback);
        },
        close() {
            ws.close();
        }
    };
};
